### ISSUE TITLE
Feature request: rebuild on module change

### ISSUE THREAD

**Issue by asakusuma** [ef4/ember-auto-import#163]
Ideally, changing an imported module should rebuild that part of the app. And should work with npm/yarn link. So one could link to an external module and iterate on the module being run in the ember app.

**Comment by samselikoff** [comment]
We just ran into this.

The way auto-import behaves today puts users into something of an "uncanny valley" – if I'm working on an Ember Component, sometimes I can add an `import` statement to the top, see my app rebuild, and have everything work, and sometimes it can't.

Importing a new module that comes from `package.json` and needs to be scraped by auto-import means my app will be in a broken state, until I quit and restart the Ember server.

We are on Ember 2.18 so maybe this behavior changed later? Is this expected behavior? If so, maybe we could force a full rebuild if/when auto-import detects a new static dependency?

**Comment by asakusuma** [comment]
@samselikoff I'm working on an ember addon at `ember-cli@3.3.0` and `ember-source@3.4.6` and I also have to restart `ember serve` anytime a dependency imported by `ember-auto-import` is updated. So I don't think the behavior has changed in 3.x.

Because of the build times of the apps I'm working on, doing an automatic full rebuild isn't going to be much of a win for me. Hopefully there is a reasonable way to just rebuild the dependencies without rebuild the entire app.

**Comment by ef4** [comment]
Sam, the thing you are reporting is different than the original bug here.
Adding an import should absolutely work without restarting. You should even
be able to npm install a dep and then add the import, without restarting
ember-cli.

This bug is about editing the auto-imported library itself, not edits to
the app.

**Comment by samselikoff** [comment]
Hm I see. Sorry for being noisy this week 😛 I will look into this and report back! We were definitely seeing it yesterday.

**Comment by austinezell** [comment]
This is a problem I am currently encountering, and would very much like to see a solution. We are developing several supporting libraries for our applications (both Ember and React apps) and using `ember-auto-import` to universally `import` them. Stopping and rerunning `ember serve` every time we develop against a local package is a drag since our Ember app is quite hefty.

Is this feature being considered/ on the roadmap?

**Comment by ef4** [comment]
I'm definitely in favor of adding this, but nobody is actively working on it at the moment.

If somebody wants to give it a shot, there are two things that need to happen. First, this:

https://github.com/ef4/ember-auto-import/blob/e36a9d8c421417c168f1254bd98c07ecb95380f0/packages/ember-auto-import/ts/bundler.ts#L92

is the line of code that prevents us from re-running webpack when the set of things you're trying to auto-import has not changed. We would want to add a new reason to return true in that conditional.

That would be enough to make changes in your library get incorporated into the next ember-cli build. But it's not enough to trigger the build to happen automatically.

So second, we'd want to make ember-cli actually watch your library's directory. It would be sufficient to make sure your directory gets added to this array:

https://github.com/ef4/ember-auto-import/blob/e36a9d8c421417c168f1254bd98c07ecb95380f0/packages/ember-auto-import/ts/bundler.ts#L42

**Comment by c-emil** [comment]
Is there any progress on this feature request? It would be awesome if anyone would implement this.

**Comment by robertsteilberg** [comment]
@ef4 I am trying to take a stab on getting a PR in to resolve this--poking around, I've removed that condition for testing purposes and ensured that webpack is getting re-run, however npm/yarn linked packages are still not updating code-wise in the ember app on rebuild. Any idea what I may be missing?

**Comment by ef4** [comment]
I don't know offhand, we would need to debug it in more detail. What I sketched about is my best understanding of the problem.

One thing to check next is whether the updated code is present in webpack's output. That would narrow down the problem.

**Comment by tomwayson** [comment]
Picking up where @robertsteilberg left off, I've cloned this repo, npm linked packages/ember-auto-import into a new Ember application, and then temporarily set this line:

https://github.com/ef4/ember-auto-import/blob/8b49e4cc293c442f9f7ff1a071b281055767829f/packages/ember-auto-import/ts/bundler.ts#L104

To  `if (bundleDeps !== this.lastDeps || linkedDepsChanged) {` where `linkedDepsChanged` is always true for now.

Then I've npm linked the library that I want to `import` from, and I've successfully been able to observe changes I make in the npm linked library by subsequently making any arbitrary change to code in the ember app that triggers a rebuild.

So, I've gotten passed this:

> however npm/yarn linked packages are still not updating code-wise in the ember app on rebuild

@ef4, I'm wondering about

> We would want to add a new reason to return true in that conditional

I've been assuming you meant we should calculate something like my `linkedDepsChanged` on the fly based on whether or not any of the files in the library's (npm linked) folder under node_modules had changed. I'm trying to figure out how we'd know if it was one of those files changed (as opposed some file in the app tree). I'm guessing that if we were to pass a the library's node_modules folder (or some subfolder) like `super([allAppTree].concat(libTrees), {` here:

https://github.com/ef4/ember-auto-import/blob/8b49e4cc293c442f9f7ff1a071b281055767829f/packages/ember-auto-import/ts/bundler.ts#L39

Where `libTrees` is an array of trees to watch based on something we'd add to `options`.

Then in `build()` we'd have access to the `libTrees` via `const [allAppPath, ...libPaths] = this.inputPaths`, but I don't see any way to tell whether it was a file in one of library paths that that triggered the rebuild other than do something like keep a checksum on each and see if that changed between builds?

Am I on the right path here?

**Comment by tomwayson** [comment]
Hmm... I stand corrected.

Now that I've got the watch wired up to my library's npm linked dist folder (i.e. `super([allAppTree].concat(libTrees)`) and I'm running webpack on any change (i.e. `if (true) { // if (bundleDeps !== this.lastDeps) { `), I've noticed that I observe the library updates in the app only _some_ of the time.

I looked into vendor.js and I've noticed that there are multiple ember-auto-import generated `eval()` statements in there for the library file that I am modifying. It appears that there is one `eval()` statement for each time I save the npm linked file, whichever one is closest to the end of vendor.js wins, and they appear to be in random order. So it becomes increasing less likely that the one that corresponds to the most recent save of the file is the last one in vendor.js and therefore reflected in the browser.

I didn't see any issues that mentioned anything like multiple `eval()` statements showing up in vendor.js. Any ideas as to whether this could be related to the changes I'm making (watching the library's npm linked dist folder and forcing webpack to run every time)?

**Comment by tomwayson** [comment]
FYI - this does not appear to be related to having set up the watch, b/c if I remove that and leave the `if (true)` to force running webpack on any change I still end up w/ an `eval()` for each time I save the file (and force the app to rebuild by subsequently saving any file in the ember app).

**Comment by tomwayson** [comment]
I've confirmed that the problem of multiple `eval()` statements for a given file exists on master (i.e. is not related to the changes on my branch at all). To replicate:

1. install 2 libraries in your app
1. use ember-auto-import to `import` from one of the libraries
1. `ember s`
1. inspect vendor.js and verify the code you imported is there (i.e. an `eval()` statement)
1. modify the imported code from that library directly in `node_modules` and save
1. now add a an `import` to the app from the second library in order to trigger another webpack build
1. inspect vendor.js and observe that there are 2 `eval()` statements for the file you modified
1. modify the the same file directly in `node_modules` again and save
1. remove the `import`  from the second library in order to trigger another webpack build
1. inspect vendor.js and observe that there are now 3 `eval()` statements for the file you modified

**Comment by tomwayson** [comment]
So, it appears _all_ the code written to vendor.js by ember-auto-import is ending up in there multiple times, not just the eval statement for the changed file! In other words, I see multiple `var __ember_auto_import__ =` in vendor.js. It seems like there's one in there for each unique combination of dependencies I `import` from. This happens even when I don't edit any files in node_modules.

For example, if my app initially imports from `package-a` there will be only one entry in vendor.js. If I later `import` from `package-b`, then I'll get a second entry in vendor.js, one has all the code for `package-a` and the other has all the code for `package-a` and `package-b`. If I then revert the app to just importing from `package-a`, I then I still both entries in vendor.js. If I then add an `import` from `package-c` I now have a third entry in vendor.js. _Now the code for `package-a` is in vendor.js 3 times!_

@ef4 I think is the result of a fundamental problem w/ how `ember-auto-import` is injecting code into vendor.js. It's only surfaced in this issue b/c we were expecting to see changes we've made to the library code, which was not always happening due to the fact that different versions were overwriting one another.

**Comment by tomwayson** [comment]
I now see [why you append the staticly imported code to vendor.js](https://github.com/ef4/ember-auto-import/issues/133#issuecomment-437582478).

I'm wondering about introducing a mode where it's not appended to vendor.js, but instead just written to a script that will be loaded after vendor.js? Developers could pass an option or env flag to enable this mode during development only. 

The only alternative I can think of is trying to parse vendor.js for the previously injected webpack output and remove it, but that seems dicey.

**Comment by tomwayson** [comment]
Actually, it looks none of that is necessary.

It seems like the issue is just that the `chunk.vendors~` files are not getting cleaned up between runs and these lines are appending the code from the files from previous runs with the file for the current run:

https://github.com/ef4/ember-auto-import/blob/8b49e4cc293c442f9f7ff1a071b281055767829f/packages/ember-auto-import/ts/broccoli-append.ts#L178-L182

So in my above example, after the initial build I have `app/chunk.vendors~app.50490d38995b2fe540b4.js` which has the webpack output with the code for `package-a`.

Then after adding the `import` from `package-b` to the app and the app rebuilds, `app/chunk.vendors~app.50490d38995b2fe540b4.js` is still there, and now `app/chunk.vendors~app.eb949ef73e7abedb13f8.js` is there, and that has the code for both `package-a` and `package-b`.

@ef4, am I right in assuming that this is not intended behavior and that we want those files cleaned between runs? If so I'll gladly make a PR that does that and will gladly take any hints of where to start.

**Comment by ef4** [comment]
Yes, it sounds like you found the bug here, thanks.

**Comment by tomwayson** [comment]
Thanks

I'm trying to PR a fix.

I've tried manually going in and cleaning up the temp dir that the output is written to before making a change that would trigger a re-build, but then I see errors like:

```
ENOENT: no such file or directory, link '/private/var/folders/_6/dgtlmj695fv5yds8f5hwvcbs8snz_f/T/broccoli-15554RTCvS8lelhoR/cache-276-bundler/output/chunk.tests.5dc7a0298b42756259d2.js' -> '/var/folders/_6/dgtlmj695fv5yds8f5hwvcbs8snz_f/T/broccoli-15554RTCvS8lelhoR/out-276-bundler/entrypoints/tests/chunk.tests.5dc7a0298b42756259d2.js'
```

There's a lot more going on the the webpack bundler and `broccoli-append` than I understand so far.

**Comment by tomwayson** [comment]
I believe @asakusuma's original request is [implemented by `watchDependencies` in 1.11+](https://github.com/ef4/ember-auto-import/issues/351#issuecomment-805455045).

@ef4, would you expect that [the bug that I identified above](https://github.com/ef4/ember-auto-import/issues/163#issuecomment-699086983) would also have been fixed as part of the `watchDependencies` work. If so I would suspect that you could close this issue.

**Comment by ef4** [comment]
Yes, this should all be covered by `watchDependencies` now.
